Update historical notices to use the new `render_notice` method.

interlineae opened the pull request:
Fixes #2378

### Changes proposed in this Pull Request

This PR updates historical licenses to use the new `WP_Job_Manager_Admin_Notices::render_notice` method.

This change changes the `render_notice` notice method from a `private static function` to a `public static function` and then updates the notice display per the renders at https://github.com/Automattic/WP-Job-Manager/issues/2382.

One item I have chosen to forgo for this change is the grouped update notification. I've decided this as the immediate request is to update historical notices to use the `render_notice` method, whereas the grouped update notifications would require the creation of new, non-historical notices.

One additional item I noticed is that the update notices don't inherit styling on non-WPJM pages:

![Screenshot from 2023-05-04 11-26-57](https://user-images.githubusercontent.com/2500940/236178802-553e46dc-1e27-43c2-a85d-6c1e1b5dd98d.png)

To address this I could either move the relevant `wp_enqueue_style` and `wp_register_style` outside of the `if ( in_array( $screen->id, apply_filters(....` conditional, or, add more screen IDs to the check. I've opted to move the style as it's not possible to anticipate all of the potential screens the notices might display on, or even if it was, the check would be so long as to be pointless.

To prevent a conflict with WC styling, I've also made use of a `! important` declaration. I don't know if, instead, we want to copy into WPJM the relevant `notice` styles and have our own `wpjm-notice` class instead. I'm not keen on `! important` so if we want to go to in that direction, I'm all for it. The declaration does seem the most straightforward though, and it's specific to our notices.

Without this change, notices lose their padding on certain WC pages (like WooCommerce --> Orders). For example:

![Screenshot from 2023-05-04 11-59-24](https://user-images.githubusercontent.com/2500940/236185464-7e9cd454-27c2-4ebb-afad-c17964ea65c0.png)

### Testing instructions

* Checkout branch
* Install a couple of paid WPJM plugins, but don't input a license key
* Notices should display with the updated formatting as shown below
* If you've tested this prior and an update notice isn't showing, use `wp option delete wp_job_manager_dismissed_notices` through WP CLI (or otherwise delete this option) to force the notice to display.

![Screenshot from 2023-05-04 11-03-16](https://user-images.githubusercontent.com/2500940/236173881-e769c390-cd8d-4580-9fb5-3d418764e9e0.png)

* To show an error notice, comment out the code to show the missing license notice https://github.com/Automattic/WP-Job-Manager/blob/15393a2e1a84d2d194f0097d2023e2940ac38e4e/includes/helper/class-wp-job-manager-helper.php#L569-L576
* change the `elseif` into an `if`, and then the check to show an error notice from `! empty` to `empty` https://github.com/Automattic/WP-Job-Manager/blob/15393a2e1a84d2d194f0097d2023e2940ac38e4e/includes/helper/class-wp-job-manager-helper.php#L576-L583

interlineae committed:
Move license key and error notices to `render_notice` method

- Makes the `render_notice` method public so that it's callable from `html-license-key-error.php and `html-license-key-notice.php`

- HTML removed from both files as this is provided as part of the `render_notice` method.

- Notices are now sent in the format of the API request the `render_notice` method expects.

interlineae committed:
Move license key and error notices to `render_notice` method

- Makes the `render_notice` method public so that it's callable from `html-license-key-error.php and `html-license-key-notice.php`

- HTML removed from both files as this is provided as part of the `render_notice` method.

- Notices are now sent in the format of the API request the `render_notice` method expects.

interlineae committed:
Load notice styling throughout WP Admin

This change loads notice styling throughout WP Admin, rather than restrict it to certain pages. This is because historical notices display throughout WP Admin and not just on WPJM pages.

WC styling was also interferring with WPJMs notice display. This has been addressed with a well-considered `! important` declaration.

interlineae commented on assets/css/admin-notices.scss:
```diff
@@ -11,7 +11,7 @@ $notice-success: #43af99;
 	// Spacing.
 	gap: 10px;
-	padding: 10px 12px;
+	padding: 10px 12px !important;
```
I'm not totally cool with this, but after testing a few other options, it's the most straightforward option and limited to WPJM notice styling.

Another option is to rename `notice` to `wpjm-notice` and copy the relevant styling over to `admin-notices.scss`.

renatho replied:
I realized an issue that we're also overriding an style that increases the padding in order to not have the dismiss button over the text.

<img width="536" alt="Screenshot 2023-05-05 at 16 28 33" src="https://user-images.githubusercontent.com/876340/236553533-945803be-b69b-48a4-a47a-c08b01b83397.png">

Maybe a solution could be to add the `!important` only on the `padding-top` and `padding-bottom`?

interlineae replied:
Ah, I see what's happening there, yeah. I've gone with the separated `padding` route and I'm not seeing any issues with that.

Committed here https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/b63ae05c8050adb24085c5ed9fdf743fca25abeb

interlineae commented on includes/admin/class-wp-job-manager-admin.php:
```diff
@@ -128,6 +125,9 @@ public function admin_enqueue_scripts() {
 		wp_enqueue_style( 'job_manager_admin_menu_css' );
+
+		WP_Job_Manager::register_style( 'job_manager_admin_notices_css', 'css/admin-notices.css', [] );
+		wp_enqueue_style( 'job_manager_admin_notices_css' );
```
I've moved this outside of the conditional above as I was adding so many different screen IDs that the check was becoming pointless. It's also not possible to anticipate IDs as users will have very many different screens based on very many different plugins, and loading the styles globally makes the most sense to me.

renatho replied:
It makes sense to me too! Just to confirm, I noticed you also removed the `dashboard` from the conditional. Was it there only for the notices?

interlineae replied:
That's right. I added that here https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/50a8a711471ebed61d76c9222f458c1524530dce.

interlineae commented:
Noting that since the feature branch has been closed by @aaronfc, approval of this PR will result in a merge directly to `trunk`, triggering a deployment !!

renatho commented on includes/helper/views/html-licence-key-error.php:
```diff
@@ -8,9 +8,18 @@
+
+$notice = [
+	'level'       => 'error',
+	'dismissible' => true,
```
I noticed that with this change, the dismissal is not being persisted anymore.

It seems we have a method checking the `wpjm_hide_notice` query string, but in this case, it's dismissing the notice without adding the query string or running any request.

Do you know if it's working in any other place to see what is different?

interlineae replied:
This was addressed in the refactor https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/15393a2e1a84d2d194f0097d2023e2940ac38e4e

renatho reviewed (commented):
Hey! In general, it's looking very good!
I just left comments about 2 things I noticed.

> Noting that since the feature branch has been closed by @aaronfc, approval of this PR will result in a merge directly to `trunk`, triggering a deployment !!

About this, I think it's not the case for the Job Manager plugin. :)

interlineae committed:
Alter notice padding to prevent conflicts

This sets the padding on the wpjm-admin-notice class to use the explicit top/right/bottom/left values. This way, the top and bottom styling can be overriden while the left/right are left alone.

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into add/update-legacy-notice-styling

interlineae committed:
Refactor notice update

- Revert `render_notice` to a `private` method
- Remove `html-licence-key-error.php` and `html-licence-key-notice.php`
- Add two new methods, `add_missing_licence_notice` and `add_error_licence_notice`
- Hook these two methods into the `wpjm_admin_notice` filter
- These can both be used as examples for adding notice to WPJM in the future

interlineae committed:
Licence and license

- Make sure 'License' follows not the UK standard

interlineae commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -545,39 +545,41 @@ function ( $plugin ) use ( $search_term ) {
-	 * Outputs unset license key notices.
+	 * Outputs unset licence key notices.
 	 */
-	public function licence_error_notices() {
-		$screen = get_current_screen();
-		if (
-			null === $screen ||
-			in_array( $screen->id, [ 'job_listing_page_job-manager-addons' ], true ) ||
-			! current_user_can( 'update_plugins' )
-		) {
-			return;
-		}
```
After speaking with @aaronfc we realised that `null` was being encountered when it ought not have been. This was removed as we decided that (1) error notices and (2) missing license notices should be displayed on the add-ons page.

interlineae commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -545,39 +545,41 @@ function ( $plugin ) use ( $search_term ) {
-					include 'views/html-licence-key-error.php';
+					add_filter(
+						'wpjm_admin_notices',
+						function ( $notices ) use ( $product_slug, $plugin_data ) {
```
I tried to avoid the anonymous functions here, but I'm not aware of another way to pass `$product_slug` and `$plugin_data` to the methods called.

aaronfc replied:
Mmm good call!

I am fine with using an anonymous function here. The idea to not use an anonymous function is to be able to "unhook" it programatically. But in this case we can always "hook" later and remove the notice_id that we want to remove.

I can think of alternative ways like storing in a private attribute the data needed or something like that, but I don't like it. Good to go with the anonymous function 🚀

interlineae commented:
@aaronfc @renatho After speaking with Aaron, I completely refactored the notice displays to use the new `wpjm_admin_notice` filter. Testing instructions have been updated, and I worked through a couple of hangups with Aaron's help. But we're good for another round of testing.

https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/15393a2e1a84d2d194f0097d2023e2940ac38e4e

Edit: oh hang on, apparently we have unit tests now 👀 
Edit: Fixed up. Failing tests now pertain to PHP version conflicts in the test suite itself

interlineae committed:
Merge branch 'trunk' into add/update-legacy-notice-styling

interlineae committed:
Fix 'licenced' leftover that was causing fatal

interlineae committed:
Merge remote-tracking branch 'origin/add/update-legacy-notice-styling' into add/update-legacy-notice-styling

interlineae committed:
Fix spelling again

I should have just reverted the commit that updated the spelling of licence, but this also fixes it. Lesson learned.

renatho commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -545,39 +545,41 @@ function ( $plugin ) use ( $search_term ) {
-	 * Outputs unset license key notices.
+	 * Outputs unset licence key notices.
 	 */
-	public function licence_error_notices() {
```
In the plugins open for the users we should take care to not remove the public functions to avoid errors with any type of integration.

It seems just a rename right? So in this case, we could deprecate this method, and inside it, call the `maybe_add_licence_error_notices`. So if someone is using it, they will see the deprecation warning instead of an error, and the function will continue working.

interlineae replied:
@aaronfc and I had discussed renaming, so it's just a rename that's right.

I've not deprecated a method before, but that sounds like a plan to me. I'll get to looking at that.

interlineae replied:
Deprecated here https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/ed89a13d0a2f23d86fb65b376bfef5dae6322a28

aaronfc replied:
Looking nice to me. Good catch, @renatho !

renatho replied:
Thank you for changing that, Katja! And sorry for the late delay on reviewing it!

Just instead of `maybe_add_license_error_notices()` we should use `WP_Job_Manager_Helper::maybe_add_license_error_notices` to reference the correct method. Not something critical and urgent though. 😉

renatho reviewed (commented):
Hi Katja!

Thank you for working on that!

In my new test, it didn't show the notices. Maybe it's related to the order of hooks where we're using the `wpjm_admin_notices` now? Let me know if you need any help with that. ;)

Also, as we talked, if we replace the "license" with "licence" in this PR, we could try to avoid the `apply_filters` and the `log_event` functions to avoid breaking filters and losing metrics. Does it make sense? I'd advocate to not rename the strings too to not lose translations of what is already translated. But it's less critical if we really want this change.

interlineae committed:
Hard reset and re-apply

This commit hard-resets to `trunk`, and then selectively applies the changes we want. Too much had crept in with the wording errancy.

interlineae committed:
Merge remote-tracking branch 'origin/add/update-legacy-notice-styling' into add/update-legacy-notice-styling

# Conflicts:
#	includes/helper/class-wp-job-manager-helper.php

interlineae committed:
Fix test spelling

interlineae commented:
> Also, as we talked, if we replace the "license" with "licence" in this PR, we could try to avoid the apply_filters and the log_event functions to avoid breaking filters and losing metrics. Does it make sense? I'd advocate to not rename the strings too to not lose translations of what is already translated. But it's less critical if we really want this change.

We did, and completely agree. I've gone ahead and hard-reset and reapplied changes:

- https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/8755fa325110855241e9e5ee8e0c4a8570696cec
- https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/aeba146ebb5702c7f49fdd8f7981ec9cb541e5be

The notices are showing for me though, so let me run through them again and see if I can replicate.

interlineae committed:
Deprecate `license_error_notices()`

- This change deprecates the `licence_error_notices()` method and calls inside it, the `maybe_add_licens_error_notices()` method. Added for compatibility in case anyone was using the method as it was.

interlineae commented:
> In my new test, it didn't show the notices. Maybe it's related to the order of hooks where we're using the `wpjm_admin_notices` now? Let me know if you need any help with that. ;)

I think this is due to previous testing. Can you delete the `wp_job_manager_dismissed_notices` option and see if that forces the notice to display anew for you? In WP CLI that would be `wp option delete wp_job_manager_dismissed_notices`.

interlineae committed:
Prefix notice with addon slug

This change adds the $product_slug to the `data-dismiss-notice` property to ensure different addons use different properties names.

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -10,14 +10,14 @@
 class WP_Job_Manager_Helper {
 	/**
-	 * License messages to display to the user.
+	 * Licence messages to display to the user.
```
I would leave this as they were (with s).

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/d9e812b90a5b4c38a01c1fb96ed29c40c2c1cad0

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -559,25 +559,39 @@ private function show_bulk_activation_form( $licenced_plugins ) {
-		) {
-			return;
-		}
+		_deprecated_function( __METHOD__, '1.33.0', 'maybe_add_license_error_notices()' );
```
Here you call the method 'license' but the real name you set is 'licence'.

I would go with 's' everywhere (except from the deprecated method, which obviously has to be kept like it was before).

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/d9e812b90a5b4c38a01c1fb96ed29c40c2c1cad0

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -559,25 +559,39 @@ private function show_bulk_activation_form( $licenced_plugins ) {
-			return;
-		}
+		_deprecated_function( __METHOD__, '1.33.0', 'maybe_add_license_error_notices()' );
+		$this->maybe_add_license_error_notices();
```
I think there is a typo here. The method is actually called "licence" (but I think you should rename it to "license").

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -895,6 +909,71 @@ private function handle_product_activation_response( $response, $product_slug, $
+	 * @param array  $plugin_data The plugin data.
+	 * @return array
+	 */
+	public function add_missing_licence_notice( $notices, $product_slug, $plugin_data ) {
```
this can be private, right?

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -895,6 +909,71 @@ private function handle_product_activation_response( $response, $product_slug, $
+	 * @param array  $plugin_data The plugin data.
+	 * @return array
+	 */
+	public function add_error_licence_notice( $notices, $product_slug, $plugin_data ) {
```
same here, can be private :)

aaronfc commented on includes/helper/class-wp-job-manager-helper.php:
```diff
@@ -895,6 +909,71 @@ private function handle_product_activation_response( $response, $product_slug, $
+				esc_html( $plugin_data['Name'] )
+			),
+		];
+		$notices[ 'wpjm_licence_notice_' . $product_slug ] = $notice;
```
I would add here "missing" to the license notice ID. Also, since the dismissing mechanism is different I think we can rename it freely (and also name it "license" with an S).

Final name: `'wpjm_missing_license_notice_' . $product_slug`

interlineae replied:
Committed (for here and above) https://github.com/Automattic/WP-Job-Manager/pull/2409/commits/ea758b00beec40385a238ad15d3a41ccaccb878a

aaronfc reviewed (approved):
Works great! 🚀 

I left some comments but are mostly about naming. Once that's fixed I am happy with this being merged.

Thanks for working on this! <3

interlineae committed:
Settle on `s` spelling

Updates the new methods, and a string, to use the preferred `s` spelling, rathern than the `c` spelling

interlineae committed:
Private rather than public class

- Make our two new helper methods private rather than public
- Change the missing-license notice attribute to `wpjm_missing_license_notice_`

interlineae committed:
Andother 'c' to 's' update

aaronfc reviewed (approved):
Looks good, works well! 🚀
